### DISCUSSION TITLE
feature idea: how about a `get_np_pixels()` method?

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#328] [Ideas]
We have `get_pixels()` that can get us a single pixel or a PImage (should it be a Py5Image?) area of pixels. Could we get a np_array quickly without having to do the `load_np_pixels()` thing before using `.np_pixels[]` etc.? Would it make sense?

I suppose there could be a performance penalty for calling load_np_pixels every time...

**Comment by hx2A** [comment]
> We have get_pixels() that can get us a single pixel or a PImage (should it be a Py5Image?) area of pixels.

Yes, it is a Py5Image object. Is there documentation somewhere that says PImage? I can't find any.

> Could we get a np_array quickly without having to do the load_np_pixels() thing before using .np_pixels[] etc.? Would it make sense?

Why not? Sounds like a great idea to me. Let's add it!

> I suppose there could be a performance penalty for calling load_np_pixels every time...

Yes, but that's already happening when users use `get_pixels()`, so it's fine.

Right now one could do the following to get a section of the pixels as a numpy array:

```python
py5.load_np_pixels()
array = py5.np_pixels[10:50, 10:50, :]
```

A new `get_np_pixels()` method would basically wrap the above into a method and return `array`. The only design question is should we add a `.copy()` at the end there. Recall that numpy will create (usually? always?) a view when you index into an array like that.

```python
In [2]: numbers = np.random.rand(5)

In [3]: numbers
Out[3]: array([0.02467264, 0.73513364, 0.8310362 , 0.90134441, 0.57971202])

In [4]: view1 = numbers[2:4]

In [5]: copy1 = numbers[2:4].copy()

In [6]: numbers[3] = 42

In [7]: view1
Out[7]: array([ 0.8310362, 42.       ])

In [8]: copy1
Out[8]: array([0.8310362 , 0.90134441])
```

I think including a `copy()` would be the correct choice, as it would also make the functionality more analogous with what `get_pixels()` does with its returned `Py5Image` object, and would be less confusing to beginner coders. There is a small performance penalty with the `copy()` but if that concerned someone they could just use `np_pixels[]` directly. All of this could be articulated in the documentation.

**Reply by villares** [reply]
Yes, the "view" thing can be a bit confusing, but is fantastic for getting performance gains!

Nevermind, I agree it would make sense for `get_np_pixels()` to be a copy.

I think I saw a "core PImage" from `get_pixels` in the interactive console, but I can't check it right now.

**Reply by hx2A** [reply]
> I think I saw a "core PImage" from get_pixels in the interactive console, but I can't check it right now.

If so, that's a bug! `core.PImage` shouldn't show up anywhere. I did some investigation with `get_pixels()` and definitely get Py5Image objects, but if there is a way to get a `PImage`, please let me know.

**Reply by villares** [reply]
![image](https://github.com/py5coding/py5generator/assets/3694604/cb2391fb-e66f-4b03-a575-ebd8265e4b09)

![image](https://github.com/py5coding/py5generator/assets/3694604/adab086f-b917-4fcb-9e39-cfc43c324c9e)

**Reply by hx2A** [reply]
The release went out last weekend. No need to use the rc1 build anymore.

https://pypi.org/project/py5/0.9.1a1/

That doesn't address the issue you raised. It seems that `get_pixels()` does not work correctly when called on a Py5Image object. I checked the code (actually a data file) and I see the mistake. Thanks for catching this!!! Much respect for your attention to detail.

Will open an issue right now.

**Reply by villares** [reply]
The RC is there because I'm on a secondary laptop, on holiday, and hadn't installed the latest release...

I think it's more luck than attention to detail or due diligence, but also, it's the result of fiddling with py5 almost everyday :)

**Reply by hx2A** [reply]
The `get_pixels()` issue is now fixed.

> I think it's more luck than attention to detail or due diligence, but also, it's the result of fiddling with py5 almost everyday :)

That also works! What is the saying, "fortune favors the prepared mind"? 

> on holiday

Enjoy your time off!
